### ISSUE TITLE
Spring boot based microservice will not re-register itself after consul agent restarted

### ISSUE THREAD

**Issue by hanct** [spring-cloud/spring-cloud-consul#727]
Issue Description: 

I manage to create a simple spring-boot based microservice that register itself with a running consul agent.  But if i close the agent, and then started it again. I notice that the microservice will not retry register itself again.  I done research online, but still could not find a solution to this.  Is this the expected behavior?   

Spring Boot: v2.5.1

**Comment by hanct** [comment]
> Issue Description:
> 
> I manage to create a simple spring-boot based microservice that register itself with a running consul agent. But if i close the agent, and then started it again. I notice that the microservice will not retry register itself again. I done research online, but still could not find a solution to this. Is this the expected behavior?


Do note that i follow this https://cloud.spring.io/spring-cloud-consul/multi/multi_spring-cloud-consul-retry.html.
By simply adding the below in my gradle.build
	implementation group: 'org.springframework.retry', name: 'spring-retry', version: '1.0.3.RELEASE'
        implementation 'org.springframework.boot:spring-boot-starter-aop',

I expect the retry to just work for my spring boot based microservice.   But its not working

**Comment by spencergibb** [comment]
What version of spring cloud are you using?

**Comment by hanct** [comment]
> What version of spring cloud are you using?

Hi, 
    version 2020.0.3

**Comment by hanct** [comment]
So what is the verdict? Does spring consul's current stable version able to support re-register itself after consul agent restarted

**Comment by hanct** [comment]
Resolved. 
need to add the below 2 properties
spring.cloud.consul.discovery.heartbeat.enabled= true 
spring.cloud.consul.discovery.heartbeat.reregister-service-on-failure=true

**Comment by dmanix** [comment]
I had a similar problem but I had to do sth more. My solution below - maybe will be helpful for someone:

Reregistration is performed when consul heartbeat request fails and ReregistrationPredicate is fullfilled. But default predicate only accepts HTTP status code 500.
In my case HTTP status code was 404 - so I needed to write custom predicate, e.g. sth like:

    @Bean
    public ReregistrationPredicate reRegistrationPredicate() {
        return e -> e.getStatusCode() >= 400;
    }
